### PR TITLE
CE-3445 Catch an event of scrolling to the first <p> and select a word

### PR DESCRIPTION
**Pull request by adamkarminski**
**Work in progress. Don't merge!**

## Links

* https://wikia-inc.atlassian.net/browse/CE-3430
* Visual context: https://drive.google.com/a/wikia-inc.com/file/d/0B_mnLiH2mWTwOWYtM09FbHdCS3c/view?usp=sharing

## Description

**This is a part of an experiment that will go live on Tue, Mar 8 and will be retired after 3 weeks on Tue, Mar 29.**

The code will be served only to 10% of traffic on EN wikis. Given the experimental nature of the code we did not invest much in performance micro-optimisation (like treeWalkers - thanks @bkoval).

## Reviewers

@bkoval @lukaszkonieczny @kamilkoterba @Grunny

### CODE REVIEW

File: front/main/app/components/article-content.js [hunk 1]
```diff
@@ -540,6 +541,33 @@ export default Ember.Component.extend(
...
 				});
+		},
+
+		insertHighlightedTextEditorDemo() {
+			const highlightedId = 'highlighted-text__demo',
+				selection = window.getSelection(),
+				range = document.createRange(),
+				$paragraphs = this.$('.article-content').children('p');
+
+			let $paragraph, plain, paragraphHtml, word, $highlightedElement;
```
Comment (Grunny) [inline front/main/app/components/article-content.js]: The scope of const and let should be as small as possible, so most of these should be moved within the `for` loop (and some could become const's as a result).

File: front/main/app/components/article-content.js [hunk 2]
```diff
@@ -540,6 +541,33 @@ export default Ember.Component.extend(
...
+
+			for (let i = 1; i <= 3; i++) {
+				$paragraph = $($paragraphs[i]);
```
Comment (Grunny) [inline front/main/app/components/article-content.js]: This should avoid the `$` global as you do below with `Ember.$` but in this case could use `const $paragraph = $paragraphs.eq(i);`

**Comment by Grunny** [comment]
Two minor comments, otherwise LGTM.

File: front/main/app/components/article-content.js [hunk 3]
```diff
@@ -540,6 +541,33 @@ export default Ember.Component.extend(
 
 					$element.wrap(wrapper);
 				});
+		},
+
+		insertHighlightedTextEditorDemo() {
+			const highlightedId = 'highlighted-text__demo',
```
Comment (bkoval) [inline front/main/app/components/article-content.js]: No I can see it's sort of a BEM-alike selector, but why it is using dashes if it's an ID? :)

File: front/main/app/components/article-content.js (hunk 1 above)
Comment (bkoval) [inline front/main/app/components/article-content.js]: +1, block scopes work well with "lets".

File: front/main/app/components/article-content.js [hunk 4]
```diff
@@ -540,6 +541,33 @@ export default Ember.Component.extend(
...
+				selection = window.getSelection(),
+				range = document.createRange(),
+				$paragraphs = this.$('.article-content').children('p');
+
+			let $paragraph, plain, paragraphHtml, word, $highlightedElement;
+
+			for (let i = 1; i <= 3; i++) {
```
Comment (bkoval) [inline front/main/app/components/article-content.js]: I'd consider moving "3" to a const with a name, this way code is self-explanatory (name of the const will tell you why it's exactly 3)

File: front/main/app/components/article-content.js (hunk 2 above)
Comment (bkoval) [inline front/main/app/components/article-content.js]: Does it mean you don't care about $paragraphs[0]?

File: front/main/app/components/article-content.js [hunk 5]
```diff
@@ -540,6 +541,33 @@ export default Ember.Component.extend(
...
+					word = plain.split(' ')[1];
+					$paragraph.html(paragraphHtml.replace(word, `<span id="${highlightedId}">${word}</span>`));
+					$highlightedElement = $paragraph.find(`#${highlightedId}`);
+
+					Ember.$(window).scrollTop($highlightedElement.offset().top - 150);
+
+					range.selectNodeContents($highlightedElement[0]);
+					selection.removeAllRanges();
+					selection.addRange(range);
+					break;
```
Comment (bkoval) [inline front/main/app/components/article-content.js]: I know this is an experiment, but using break is just bad :) I would consider $(selector).slice(0,numberOfResults) to get your desired number of results. Then, use Array.some() to iterate over it and return false when the "if" condition is met. How about that? :)

**Comment by bkoval** [comment]
LGTM :hamburger: :hammer:

**Comment by kamilkoterba** [comment]
Will be tested on feature branch #2067

**Pull request merged**